### DISCUSSION TITLE
segment fault when running inside an armv7 container on x86 host

### DISCUSSION THREAD

**Discussion by dwebfan** [libvips/libvips#2099] [Q&A]
**Describe the bug**
I built one arm libvips library based on balena base image container on x86 host. Then when I run my golang application to save jpeg thumbnail, it got below segment fault.
```
(govips:957): GLib-GObject-WARNING **: 08:45:49.205: ../../../gobject/gsignal.c:2523: signal 'invalidate
' is invalid for instance '0xe9701198' of type 'VipsSequential'
qemu: uncaught target signal 11 (Segmentation fault) - core dumped
```
Then I manually enable to debug in cache.c, and got this result:
```
...
 VipsLineCache (0xe9711978) count=2
vips_cache_insert: start insert to cache
vips_cache_insert: finish insert to cache
vips_cache_ref: linecache in=((VipsImage*) 0x279c70) out=((VipsImage*) 0x294aa8) tile-height=1 access=((VipsAccess) VIPS_ACCESS_SEQUENTIAL) persistent=TRUE -
vips_cache_ref: g_object_ref
vips_cache_ref: argument_map
vips_cache_ref: touch operation
vips_cache_ref: end
vips_cache_insert: signal invalidate
vips_cache_insert: end
vips_cache_operation_add: start trimming
vips_cache_trim: trimming linecache in=((VipsImage*) 0x279c70) out=((VipsImage*) 0x294aa8) tile-height=1 access=((VipsAccess) VIPS_ACCESS_SEQUENTIAL) persistent=TRUE -
vips_cache_trim: cache remove start
 vips_cache_remove: linecache in=((VipsImage*) 0x279c70) out=((VipsImage*) 0x294aa8) tile-height=1 access=((VipsAccess) VIPS_ACCESS_SEQUENTIAL) persistent=TRUE -
vips_cache_unref: linecache in=((VipsImage*) 0x279c70) out=((VipsImage*) 0x294aa8) tile-height=1 access=((VipsAccess) VIPS_ACCESS_SEQUENTIAL) persistent=TRUE -
vips_cache_trim: cache remove done
 vips_cache_operation_add: end trimming
vips_cache_operation_add: adding hash = 648921
sequential args:
   in: offset = 64 required construct assigned
   nickname: offset = 24 set-once assigned
   trace: offset = 76 construct
   out: offset = 60 required construct assigned
   description: offset = 28 set-once assigned
   tile-height: offset = 68 construct
   access: offset = 72 construct
 VipsSequential (0xe9701198) count=2
vips_cache_insert: adding to cachehash = 648921
sequential args:
   in: offset = 64 required construct assigned
   nickname: offset = 24 set-once assigned
   trace: offset = 76 construct
   out: offset = 60 required construct assigned
   description: offset = 28 set-once assigned
   tile-height: offset = 68 c
```
Note that I added some own print function for step through debug purpose because I can not run gdb inside the container.

**To Reproduce**
It can be reproduced only inside the container, but not in target arm SBC, such as pi. Note that sometimes it will still print out `(govips:957): GLib-GObject-WARNING **:` warning message

**Expected behavior**
No segment fault.

**Actual behavior**
Segment fault.

**Screenshots**
n/a

**Environment**
(please complete the following information)
 - OS: Debian buster / armv7
 - Vips: 8.10.5

**Additional context**
n/a

**Comment by lovell** [comment]
> because I can not run gdb inside the container

This would suggest the correct privileges have not been granted to the container, which is also a possible reason for the segfault.

Please can you provide a complete, minimal, standalone repo with the relevant container setup e.g. as a Dockerfile, plus any QEMU configuration/scripts to allow someone else to reproduce the problem.

**Comment by dwebfan** [comment]
Thanks for prompt reply. Please see this repo https://github.com/lomorage/vips-docker-build. It includes one simple go program to do thumbnail generation and one folder with some test images. 

The steps is 
1. make build-vips
2. make test-vips

Then it may have this error messages. Please kindly review the codes and see if there are any issues.
```
2021/02/06 07:49:43 Vips started with concurrency=1 cache_max_files=0 cache_max_mem=0 cache_max=0
2021/02/06 07:49:43 Registered image type loader type=jpeg
2021/02/06 07:49:43 Registered image type loader type=magick
2021/02/06 07:49:43 Registered image type loader type=pdf
2021/02/06 07:49:43 Registered image type loader type=png
2021/02/06 07:49:43 Registered image type loader type=svg
2021/02/06 07:49:43 Registered image type loader type=tiff
2021/02/06 07:49:43 Registered image type loader type=webp
2021/02/06 07:49:43 Registered image type loader type=gif
start save jpeg: test-images/1.webp_480_320_thumbnail.jpg
2021/02/06 07:49:44 finish save jpeg: test-images/1.webp_480_320_thumbnail.jpg
start save jpeg: test-images/1.webp_75_75_thumbnail.jpg
2021/02/06 07:49:44 finish save jpeg: test-images/1.webp_75_75_thumbnail.jpg
start save jpeg: test-images/1.webp_320_0_thumbnail.jpg
2021/02/06 07:49:44 finish save jpeg: test-images/1.webp_320_0_thumbnail.jpg
start save jpeg: test-images/20180330_3.heic_480_320_thumbnail.jpg
qemu: uncaught target signal 11 (Segmentation fault) - core dumped
```

**Comment by lovell** [comment]
Thanks, are you able to reproduce this by calling vips at the command line within the same container? This would remove govips from the equation to see if that is a possible cause. It also looks like you're using an outdated, unmaintained fork of govips.

**Comment by dwebfan** [comment]
> Thanks, are you able to reproduce this by calling vips at the command line within the same container? This would remove govips from the equation to see if that is a possible cause. It also looks like you're using an outdated, unmaintained fork of govips.

Is vipsthumbnail support a list of files? This issue seems happened only during processing a batch of files. If I run the go program to process the image one by one, it won't have the issue either.

Also I tried the latest go packages, result is still same. The underneath function call isn't changing too much btw the two packages. I lost the change using the latest package, so just post the old one.
I haven't written c program for some time, but if needed it to debug, I can convert go program to c and try to reprodue. Please let me know.

**Comment by dwebfan** [comment]
seems like vipsthumbnail can process multiple photos, and doesn't have core dump. Let me dig more on it and get back to you.

**Comment by dwebfan** [comment]
Can not reproduce using native codes, so close the issue.

**Comment by dwebfan** [comment]
seems like this sample native program https://github.com/lomorage/vips-docker-build/blob/master/test-vips-c/main.c can reproduce the issue. Can you please help review the codes?

In case you want to reproduce, you can run as below.
1. make build-vips
2. make test-vips-c

Please let me know if you see any issues. Thanks for looking into.

**Comment by jcupitt** [comment]
I'll read the code, thanks for making the test case.
